### PR TITLE
[App] Add ready property to the flow

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces the `ready` property on the flow to enable Open UI only once certain conditions are met.

```python
class FlowReady(LightningFlow):
    def __init__(self):
        super().__init__()
        self.w = WorkReady()

    @property
    def ready(self) -> bool:
        return self.w.has_succeeded

    def run(self):
        self.w.run()

        if self.ready:
            self._exit()
```

Example with the Boring App:

```python
    @property
    def ready(self) -> bool:
        if "dst_w" in self.dict:
            return self.dict["dst_w"].url != ""
        return False
```

https://user-images.githubusercontent.com/12861981/205928911-aadc0f12-4de8-4dba-b0cb-1c8e9a742243.mov

Fixes #\<issue_number>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]
Nice! LGTM 😃

File: src/lightning_app/core/flow.py [hunk 1]
```diff
@@ -230,6 +230,12 @@ def __getattr__(self, item):
             return Path.from_dict(self._paths[item])
         return self.__getattribute__(item)
 
+    @property
+    def ready(self) -> bool:
+        """Override to customize when your App should be ready."""
```
Comment (awaelchli) [inline src/lightning_app/core/flow.py:235]: The user doesn't know what it means for it to be ready. The impact of this is that the button on the admin UI becomes enabled and the user can open the UI when the app is ready. I think we should explicitly say that here. 

Also, as discussed offline, the behavior here is special. The ready property will only be re-evaluated until it turns to True and then never again. Meaning the state here can't change arbitrarily during runtime. This probably makes sense, so I suggest to document the behavior.

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/core/app.py [hunk 2]
```diff
@@ -142,6 +142,9 @@ def __init__(
         self.exception = None
         self.collect_changes: bool = True
 
+        # TODO: Enable ready locally for opening the UI.
+        self.ready = False
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning_app/core/app.py:146]: I would prefer to do this another PR to deliver faster.

File: src/lightning_app/core/flow.py (hunk 1 above)
Comment (tchaton) [reply src/lightning_app/core/flow.py:235]: This would be clarified once local dev is added properly.

**Pull request merged**